File: cpp/src/hand_landmarks.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include "mediapipe/framework/formats/landmark.pb.h"
#include "mediapipe/framework/formats/landmark.pb.h"
#include "mediapipe/framework/formats/landmark.pb.h"
#include "mediapipe/framework/port/parse_text_proto.h"
#include "mediapipe/framework/port/file_helpers.h"
#include "mediapipe/framework/port/status.h"
#include "mediapipe/solutions/hands.h"

namespace mp = mediapipe;

int main(int argc, char** argv) {
    // Initialize MediaPipe Hands solution.
    mp::Hands hands;
    hands.Initialize(true, 0.3); // Static mode, min detection confidence

    // Load the image using OpenCV.
    cv::Mat image = cv::imread("demo_img/hand.jpg");
    if (image.empty()) {
        std::cerr << "Could not read the image." << std::endl;
        return -1;
    }

    // Convert the image to RGB as required by MediaPipe.
    cv::Mat image_rgb;
    cv::cvtColor(image, image_rgb, cv::COLOR_BGR2RGB);

    // Process the image to find hand landmarks.
    auto results = hands.Process(image_rgb);

    if (!results.multi_hand_landmarks.empty()) {
        for (const auto& hand_landmarks : results.multi_hand_landmarks) {
            // Draw the landmarks on the image using OpenCV.
            for (int i = 0; i < hand_landmarks.landmark_size(); ++i) {
                const auto& lm = hand_landmarks.landmark(i);
                cv::circle(image, cv::Point(lm.x() * image.cols, lm.y() * image.rows), 5, cv::Scalar(0, 255, 0), -1);
            }

            // Print out the landmarks.
            for (int i = 0; i < hand_landmarks.landmark_size(); ++i) {
                const auto& lm = hand_landmarks.landmark(i);
                std::cout << "Landmark " << i << ": (" << lm.x() << ", " << lm.y() << ", " << lm.z() << ")" << std::endl;
            }
        }
    }

    // Show the image with landmarks drawn on it.
    cv::imshow("Hand Landmarks", image);
    cv::waitKey(0);
    cv::destroyAllWindows();

    return 0;
}

File: rpi4/gesture_recognizer_py/hand_landmarks.py
import mediapipe as mp
import cv2
import numpy as np

# inspired from https://github.com/computervisioneng/sign-language-detector-python/blob/master/create_dataset.py

mp_hands = mp.solutions.hands
mp_drawing = mp.solutions.drawing_utils
mp_drawing_styles = mp.solutions.drawing_styles

hands = mp_hands.Hands(static_image_mode=True, min_detection_confidence=0.3)

image = cv2.imread('demo_img/hand.jpg')
image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

results = hands.process(image_rgb)

if results.multi_hand_landmarks:
    for hand_landmarks in results.multi_hand_landmarks:
        mp_drawing.draw_landmarks(
            image,
            hand_landmarks,
            mp_hands.HAND_CONNECTIONS,
            mp_drawing_styles.get_default_hand_landmarks_style(),
            mp_drawing_styles.get_default_hand_connections_style()
        )
        landmarks = np.array(
            [[lm.x, lm.y, lm.z] for lm in hand_landmarks.landmark]
        )
        print(landmarks)

cv2.imshow('Hand Landmarks', image)
cv2.waitKey(0)
cv2.destroyAllWindows()
